### PR TITLE
[zh] localize blog 2020-09-03-warnings

### LINKED ISSUE #37744: [zh-cn] Warning: Helpful Warnings Ahead doesn't has a zh-cn version
link: https://kubernetes.io/blog/2020/09/03/warnings/

### PR DESCRIPTION
**Pull request by niuzhenguo**
Localize content/zh-cn/blog/_posts/2020-09-03-warnings/index.md

Fixes #37744

### CODE REVIEW

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+
+<!--
+Along with the API server ability to warn about deprecated API use, starting in v1.19, a CustomResourceDefinition can indicate a 
+[particular version of the resource it defines is deprecated](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation).
+When API requests to a deprecated version of a custom resource are made, a warning message is returned, matching the behavior of built-in APIs.
+
+The author of the CustomResourceDefinition can also customize the warning for each version if they want to.
+This allows them to give a pointer to a migration guide or other information if needed.
+-->
+伴随着 API 服务器对已弃用的 API 使用发出警告的能力外，从 v1.19 开始，CustomResourceDefinition
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:299]:
```suggestion
除了 API 服务器对已弃用的 API 使用发出警告的能力外，从 v1.19 开始，CustomResourceDefinition
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+可以指示[它定义的资源的特定版本已弃用](/zh-cn/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation)。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:300]:
```suggestion
可以指示[它定义的资源的特定版本已被弃用](/zh-cn/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation)。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+当对自定义资源的已弃用的版本发出 API 请求时，将返回一条警告消息，与内置 API 的行为相匹配。
+
+CustomResourceDefinition 的作者还可以根据需要自定义每个版本的警告。
+这允许他们在需要时提供指向迁移指南或其他信息。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:304]:
```suggestion
这允许他们在需要时提供指向迁移指南的信息或其他信息。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+  name: crontabs.example.com
+spec:
+  versions:
+  - name: v1alpha1
+    # This indicates the v1alpha1 version of the custom resource is deprecated.
+    # API requests to this version receive a warning in the server response.
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:314]: 请翻译这里的注释

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+[Admission webhooks](/docs/reference/access-authn-authz/extensible-admission-controllers)
+are the primary way to integrate custom policies or validation with Kubernetes.
+Starting in v1.19, admission webhooks can [return warning messages](/docs/reference/access-authn-authz/extensible-admission-controllers/#response)
+that are passed along to the requesting API client. Warnings can be returned with allowed or rejected admission responses.
+
+As an example, to allow a request but warn about a configuration known not to work well, an admission webhook could send this response:
+-->
+[准入 Webhook](/zh-cn/docs/reference/access-authn-authz/extensible-admission-controllers)是将自定义策略或验证与
+Kubernetes 集成的主要方式。
+从 v1.19 开始，Admission Webhook 可以[返回警告消息](/zh-cn/docs/reference/access-authn-authz/extensible-admission-controllers/#response)
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:346]:
```suggestion
从 v1.19 开始，Admission Webhook 可以[返回警告消息](/zh-cn/docs/reference/access-authn-authz/extensible-admission-controllers/#response)，
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+传递给发送请求的 API 客户端。警告可以与允许或拒绝的响应一起返回。
+
+例如，允许请求但警告某个已知的无法正常运行的配置，准入 Webhook 可以发送以下响应：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:349]:
```suggestion
例如，允许请求但警告已知某个配置无法正常运行时，准入 Webhook 可以发送以下响应：
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+```
+
+<!--
+If you are implementing a webhook that returns a warning message, here are some tips:
+
+* Don't include a "Warning:" prefix in the message (that is added by clients on output)
+* Use warning messages to describe problems the client making the API request should correct or be aware of
+* Be brief; limit warnings to 120 characters if possible
+-->
+如果您正在实现一个返回警告消息的 Webhook，这里有一些提示：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:372]:
```suggestion
如果你在实现一个返回警告消息的 Webhook，这里有一些提示：
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+
+* 不要在消息中包含 “Warning:” 前缀（由客户端在输出时添加）
+* 使用警告消息来正确描述能被发出 API 请求的客户端识别的问题
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:375]:
```suggestion
* 使用警告消息来正确描述能被发出 API 请求的客户端纠正或了解的问题
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+* 保持简洁；如果可能，将警告限制为120个字符
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:376]:
```suggestion
* 保持简洁；如果可能，将警告限制为 120 个字符以内
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+
+<!--
+There are many ways admission webhooks could use this new feature, and I'm looking forward to seeing what people come up with.
+Here are a couple ideas to get you started:
+
+* webhook implementations adding a "complain" mode, where they return warnings instead of rejections,
+  to allow trying out a policy to verify it is working as expected before starting to enforce it
+* "lint" or "vet"-style webhooks, inspecting objects and surfacing warnings when best practices are not followed
+-->
+准入 Webhook 可以通过多种方式使用这个新功能，我期待看到大家想出来。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:386]:
```suggestion
准入 Webhook 可以通过多种方式使用这个新功能，我期待看到大家想出来的方法。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+
+<!--
+Applications that use the `k8s.io/client-go` library to make API requests can customize
+how warnings returned from the server are handled. By default, warnings are logged to
+stderr as they are received, but this behavior can be customized 
+[per-process](https://godoc.org/k8s.io/client-go/rest#SetDefaultWarningHandler)
+or [per-client](https://godoc.org/k8s.io/client-go/rest#Config).
+-->
+使用 `k8s.io/client-go` 库发出 API 请求的应用程序可以自定义
+如何处理从服务器返回的警告。默认情况下，收到的告警会记录到 stderr，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:406]:
```suggestion
使用 `k8s.io/client-go` 库发出 API 请求的应用程序可以定制如何处理从服务器返回的警告。
默认情况下，收到的告警会以日志形式输出到 stderr，
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+但[每个进程](https://godoc.org/k8s.io/client-go/rest#SetDefaultWarningHandler)
+或 [每个客户端](https://godoc.org/k8s.io/client-go/rest#Config)的这个行为可以自定义。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:408]:
```suggestion
但[在进程层面](https://godoc.org/k8s.io/client-go/rest#SetDefaultWarningHandler)或[客户端层面](https://godoc.org/k8s.io/client-go/rest#Config)均可定制这一行为。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+
+<!--
+This example shows how to make your application behave like `kubectl`,
+overriding message handling process-wide to deduplicate warnings 
+and highlighting messages using colored output where supported:
+-->
+这个例子展示了如何让你的应用程序表现得像 `kubectl`，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:415]:
```suggestion
下面的例子展示如何让你的应用程序表现得像 `kubectl`，
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+覆盖整个消息处理过程以删除重复警告，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:416]:
```suggestion
在进程层面重载整个消息处理逻辑以删除重复的警告，
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+```
+
+<!--
+The next example shows how to construct a client that ignores warnings.
+This is useful for clients that operate on metadata for all resource types
+(found dynamically at runtime using the discovery API)
+and do not benefit from warnings about a particular resource being deprecated.
+Suppressing deprecation warnings is not recommended for clients that require use of particular APIs.
+-->
+下一个示例展示了如何构建一个忽略告警的客户端。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:448]:
```suggestion
下一个示例展示如何构建一个忽略告警的客户端。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 16]
```diff
@@ -0,0 +1,513 @@
...
+这对于那些操作所有资源类型（使用发现 API 在运行时动态发现）的元数据
+并且不会从已弃用的特定资源的警告中受益的客户端很有用。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:450]:
```suggestion
这对于那些操作所有资源类型（使用发现 API 在运行时动态发现）
的元数据并且不会从已弃用的特定资源的警告上受益的客户端很有用。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 17]
```diff
@@ -0,0 +1,513 @@
...
+## Kubectl Strict Mode
+-->
+## Kubectl 强制模式    {#kubectl-strict-mode}
+
+<!--
+If you want to be sure you notice deprecations as soon as possible and get a jump start on addressing them,
+`kubectl` added a `--warnings-as-errors` option in v1.19. When invoked with this option,
+`kubectl` treats any warnings it receives from the server as errors and exits with a non-zero exit code:
+-->
+如果您想确保尽快发现到弃用并立即着手解决它们，
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:479]:
```suggestion
如果你想确保及时注意到弃用问题并立即着手解决它们，
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 18]
```diff
@@ -0,0 +1,513 @@
...
+`kubectl` 在 v1.19 中添加了 `--warnings-as-errors` 选项。使用此选项调用时，
+`kubectl` 将从服务器收到的任何告警视为错误并以非零码退出：
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:481]:
```suggestion
`kubectl` 将从服务器收到的所有告警视为错误，并以非零码退出：
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 19]
```diff
@@ -0,0 +1,513 @@
...
+
+<img alt="kubectl applying a manifest file with a --warnings-as-errors flag, displaying a warning message and exiting with a non-zero exit code."
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:483]: alt 可以翻译

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 20]
```diff
@@ -0,0 +1,513 @@
...
+     src="kubectl-warnings-as-errors.png"
+     style="width:637px;max-width:100%;">
+
+<!--
+This could be used in a CI job to apply manifests to a current server,
+and required to pass with a zero exit code in order for the CI job to succeed.
+-->
+这可以在 CI 作业中用于将清单文件应用到当前服务器，
+并且要求通过零退出码才能使 CI 作业成功。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:492]:
```suggestion
其中要求通过零退出码才能使 CI 作业成功。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 21]
```diff
@@ -0,0 +1,513 @@
...
+Now that we have a way to communicate helpful information to users in context,
+we're already considering other ways we can use this to improve people's experience with Kubernetes.
+A couple areas we're looking at next are warning about [known problematic values](http://issue.k8s.io/64841#issuecomment-395141013)
+we cannot reject outright for compatibility reasons, and warning about use of deprecated fields or field values
+(like selectors using beta os/arch node labels, [deprecated in v1.14](/docs/reference/labels-annotations-taints/#beta-kubernetes-io-arch-deprecated)).
+I'm excited to see progress in this area, continuing to make it easier to use Kubernetes.
+-->
+现在我们有了一种在上下文中向用户传达有用信息的方法，
+我们已经在考虑使用其他方法来改善人们使用 Kubernetes 的体验。
+我们接下来要研究的几个领域是关于[已知有问题的值](http://issue.k8s.io/64841#issuecomment-395141013)的告警
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:509]:
```suggestion
我们接下来要研究的几个领域是关于[已知有问题的值](http://issue.k8s.io/64841#issuecomment-395141013)的告警。
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 22]
```diff
@@ -0,0 +1,513 @@
...
+出于兼容性原因，我们不能直接拒绝，并警告使用已弃用的字段或字段值
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:510]:
```suggestion
出于兼容性原因，我们不能直接拒绝，而应就使用已弃用的字段或字段值
```

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 23]
```diff
@@ -0,0 +1,513 @@
...
+（例如使用 beta os/arch 节点标签的选择器，
+[在 v1.14 中已弃用](/zh-cn/docs/reference/labels-annotations-taints/#beta-kubernetes-io-arch-deprecated)）。
```
Comment (tengqm) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:512]:
```suggestion
[在 v1.14 中已弃用](/zh-cn/docs/reference/labels-annotations-taints/#beta-kubernetes-io-arch-deprecated)）
给出警告。
```

**Review by windsonsea — COMMENTED** [review]
When you search `./content/zh-cn`, you can get 11 `告警` and 121 `警告`, so `警告` is preferred. To keep consistency on a page can reduce confusion to readers.

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 24]
```diff
@@ -0,0 +1,555 @@
+---
+layout: blog
+title: "告警: 有用的预警"
```
[the remark this answers is not in the document]
Comment (niuzhenguo) [reply content/zh-cn/blog/_posts/2020-09-03-warnings/index.md]: done

**Comment by niuzhenguo** [comment]
@tengqm @windsonsea already updated according to your comments, pls have a look at this when you got time, thanks!

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md [hunk 25]
```diff
@@ -0,0 +1,555 @@
...
+```
+
+<!--
+If you are implementing a webhook that returns a warning message, here are some tips:
+
+* Don't include a "Warning:" prefix in the message (that is added by clients on output)
+* Use warning messages to describe problems the client making the API request should correct or be aware of
+* Be brief; limit warnings to 120 characters if possible
+-->
+如果您在实现一个返回警告消息的 Webhook，这里有一些提示：
```
Comment (windsonsea) [inline content/zh-cn/blog/_posts/2020-09-03-warnings/index.md:410]:
```suggestion
如果你在实现一个返回警告消息的 Webhook，这里有一些提示：
```

**Review by windsonsea — COMMENTED** [review]
Take it easy. Reviewers will come when they are available. This PR is relatively big and need some time to think over and check details.

File: content/zh-cn/blog/_posts/2020-09-03-warnings/index.md (hunk 25 above)
Comment (niuzhenguo) [reply content/zh-cn/blog/_posts/2020-09-03-warnings/index.md]: done

**Pull request merged**